### PR TITLE
add ability to pick persisted case tile from another module

### PR DESCRIPTION
**Pull request by mkangia**
extracted from #16477 
[trello card](https://trello.com/c/wekRPsTb/137-app-builder-changes-for-enikshay-phase-2b?menu=filter)

This adds ability to use a different module for
1. case tiles in case list
2. case tiles persisted over the forms

🏛

### CODE REVIEW

File: corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html [hunk 1]
```diff
@@ -56,6 +58,8 @@
     </div>
     {% endif %}
 
+    {{ request|toggle_tag_info:"CASE_LIST_TILE" }}
+    {% if request|toggle_enabled:'CASE_LIST_TILE' %}
```
Comment (snopoke) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html]: how come this is going behind a toggle now? Seems like it's already behind the `CASE_LIST_CUSTOM_XML` toggle.

File: corehq/feature_previews.py [hunk 2]
```diff
@@ -127,3 +127,12 @@ def enable_callcenter(domain_name, checked):
     ),
     privilege=LOOKUP_TABLES,
 )
+
+
+SHARE_CASE_TILE_CONTEXT = FeaturePreview(
```
Comment (snopoke) [inline corehq/feature_previews.py]: Don't think this should be a feature preview since there are other toggles that are required to make this work.

Comment (snopoke) [inline corehq/feature_previews.py]: In fact since this is basically an extension to the case tiles feature I don't think it needs a separate toggle at all and can just piggy back on the `CASE_LIST_TILE` toggle.

File: corehq/apps/app_manager/templatetags/app_manager_extras.py [hunk 3]
```diff
@@ -16,3 +16,9 @@ def get_available_modules_for_case_list_configuration(app, module):
             and m.module_type not in disallowed_module_types
             and m.case_type == module.case_type)
     ]
+
+
+@register.filter
+def get_available_modules_for_case_tile_configuration(app, module):
```
Comment (snopoke) [inline corehq/apps/app_manager/templatetags/app_manager_extras.py]: rename `module` to `exclude_module`

File: corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html [hunk 4]
```diff
@@ -76,6 +79,34 @@ <h4 class="panel-title panel-title-nolink">
                 </label>
             </div>
         </div>
+        {% endif %}
+
+        {% if request|feature_preview_enabled:"SHARE_CASE_TILE_CONTEXT" %}
+        {{ request|toggle_tag_info:"CASE_LIST_TILE" }}
+        {% if request|toggle_enabled:'CASE_LIST_TILE' %}
+        {% if request|toggle_enabled:'SHOW_PERSIST_CASE_CONTEXT_SETTING' %}
```
Comment (snopoke) [inline corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html]: having something under 2 toggles seems pretty weird. Can we not clean up the toggles so that there's only 1?

**Review by snopoke — CHANGES_REQUESTED** [review]

File: corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html (hunk 1 above)
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html]: the `CASE_LIST_TILE` toggle allows user to set `useCaseTiles` property. If someone sets that to 'yes' and then the toggle is removed for the domain, the container below is still visible since `useCaseTiles` was set to `yes`.

Oh! i see. this now makes it unavailable in case just `CASE_LIST_CUSTOM_XML` toggle is enabled. Would need to add check for that toggle above instead of `data-bind` attr.

File: corehq/apps/app_manager/templatetags/app_manager_extras.py (hunk 3 above)
Comment (mkangia) [inline corehq/apps/app_manager/templatetags/app_manager_extras.py]: using `for_module` sounds better?

Comment (snopoke) [inline corehq/apps/app_manager/templatetags/app_manager_extras.py]: if you read the function as a sentence "get available modules for case tile configuration [for] app [but exclude this] module"

I don't know not really that important

File: corehq/feature_previews.py (hunk 2 above)
Comment (mkangia) [inline corehq/feature_previews.py]: i am not completely sure about how feature previews are related to toggles, but it seems alright to move this under `SHOW_PERSIST_CASE_CONTEXT_SETTING` since we are talking about persistent case tiles.

File: corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html (hunk 4 above)
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html]: i am not sure if that would be good to do. Those two toggles might be used by different set of domains.

Though just the check for `SHOW_PERSIST_CASE_CONTEXT_SETTING` is sufficient but if somebody enables that without enabling `CASE_LIST_TILE`, this might give unexpected results in case the case tiles are configured and then the `CASE_LIST_TILE` was removed.

File: corehq/feature_previews.py (hunk 2 above)
Comment (mkangia) [inline corehq/feature_previews.py]: also `SHOW_PERSIST_CASE_CONTEXT_SETTING` seems like being used only by a couple of active real domains and a bunch of test domains so this sounds safe enough to move under `SHOW_PERSIST_CASE_CONTEXT_SETTING` only.

File: corehq/apps/app_manager/templatetags/app_manager_extras.py (hunk 3 above)
Comment (mkangia) [inline corehq/apps/app_manager/templatetags/app_manager_extras.py]: i read this as [for] app [for] module. 
The logic is not just about getting all modules except this one.  There is a method above that already does that. This method gives the module that are valid options for the configuration.

Comment (snopoke) [inline corehq/apps/app_manager/templatetags/app_manager_extras.py]: Yes but the only reason you pass in the current module is so that it doesn't get included in the returned list. Anyway, not worth wasting airtime on :)

**Review by benrudolph — APPROVED** [review]
don't have a lot of context on this change, but the code looks good (looks like still one comment pending). toggle soup! 🍚

**Comment by mkangia** [comment]
@snopoke sorry misinterpreted `SHOW_PERSIST_CASE_CONTEXT_SETTING`. Please don't review further yet. Need to fix this.

File: corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html (hunk 1 above)
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v1/partials/case_list.html]: Fixed this by moving the toggle check in the data-bind attr.

File: corehq/feature_previews.py (hunk 2 above)
Comment (mkangia) [inline corehq/feature_previews.py]: Removed Feature Preview.

File: corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html (hunk 4 above)
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html]: check for `SHOW_PERSIST_CASE_CONTEXT_SETTING` was not needed. I misinterpreted the context of that toggle. Removed it now.

**Comment by mkangia** [comment]
@snopoke fixed now. Good to review.

File: corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html [hunk 5]
```diff
@@ -62,7 +63,7 @@ <h4 class="panel-title panel-title-nolink">
             </div>
         </div>
         {% endif %}
-        <div data-bind="visible: useCaseTiles() == 'yes' || COMMCAREHQ.toggleEnabled('CASE_LIST_CUSTOM_XML')">
+        <div data-bind="visible: (useCaseTiles() == 'yes' && COMMCAREHQ.toggleEnabled('CASE_LIST_TILE')) || COMMCAREHQ.toggleEnabled('CASE_LIST_CUSTOM_XML')">
```
Comment (snopoke) [inline corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html]: Why is this change necessary? Changing this doesn't seem like part of the feature you're adding and I think the original logic is fine.

**Review by snopoke — COMMENTED** [review]
Looks good. Just comment on toggle logic.

File: corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html (hunk 5 above)
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v2/partials/case_list.html]: Yup! its not a part of the feature that i am adding but an improvement in what is present

the reason
```
the CASE_LIST_TILE toggle allows user to set useCaseTiles property. If someone sets that to 'yes' and then the toggle is removed for the domain, the container below is still visible since useCaseTiles was set to yes.
```

**Comment by snopoke** [comment]
Yes that's correct and I think that's how it should be since it is still
set in the app and hence will still appear on the phone. If we hide it when
the toggle is removed then there is no way to remove it from the app.

Simon Kelly
Director of Server Engineer | Dimagi

**Comment by mkangia** [comment]
@snopoke cool. Removing that change.

**Review by snopoke — APPROVED** [review]
